lib/resourcemerge/batch.go

```diff
@@ -8,21 +8,26 @@ import (
 // EnsureJob ensures that the existing matches the required.
 // modified is set to true when existing had to be updated with required.
 func EnsureJob(modified *bool, existing *batchv1.Job, required batchv1.Job) {
-	EnsureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
 
-	if existing.Spec.Selector == nil {
-		*modified = true
-		existing.Spec.Selector = required.Spec.Selector
-	}
-	if !equality.Semantic.DeepEqual(existing.Spec.Selector, required.Spec.Selector) {
-		*modified = true
-		existing.Spec.Selector = required.Spec.Selector
+	// To set Selector you must set ManualSelector to True. If 'required'
+	// sets Selector to value not equal to existing, let current Job
+	// finish as Selector is immutable.
+	if required.Spec.ManualSelector != nil {
+
+		// If ManualSelector is not True we don't care about Selector
+		if *required.Spec.ManualSelector {
```

Comment:
Shifting the `ManualSelector` `setBoolPtr` back to its original location will allow you to collapse these cases into a single:

```go
if required.Spec.ManualSelector != nil &&
   *required.Spec.ManualSelector &&
   !equality.Semantic.DeepEqual(existing.Spec.Selector, required.Spec.Selector) {
```

Reply:
Talked to @jottofar out of band, and he linked to [these][1] [docs][2].  After reading them over, it seems like setting your own Job `selector` is useful for fancy shell games where you delete/replace Jobs without interrupting the underlying Pods.  It's not clear to me that CVO consumers would need anything that sophisticated, and certainly not clear to me how they would know what to feed into `selector` in their static manifest even if they did want something like that.  So how about:

```go
if required.Spec.Selector != nil {
  panic("spec.selector is not supported in Job manifests")
}
```

And ideally we'd reroll the resource managers to allow for returning errors instead of panicking, but a refactor like that is probably too invasive to be backportable, so we shouldn't do it in this PR.

[1]: https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#specifying-your-own-pod-selector
[2]: https://cloud.google.com/kubernetes-engine/docs/how-to/jobs#specifying_a_pod_selector

Reply:
True, if we move setBoolPtr. I find it more logical to have steBoolPtr where I moved it but could still do a collapse. My main goal here is to first ensure the logic is correct.